Crypto error handling

martinthomson opened the pull request:
Alex pointed out a few potential bugs in the code.  This fixes the integer overflow and the UAF, but in doing the latter I decided that it would be better to centralize the error handling into the experimental API macro.  That allows the wrapper to report errors with CString::new() rather than just calling panic!.

Fixes #145.

agrover commented on neqo-crypto/src/result.rs:
```diff
@@ -34,16 +25,12 @@ where
 }
 
-fn result_helper(rv: ssl::SECStatus, allow_blocked: bool) -> Res<bool> {
+pub fn result(rv: ssl::SECStatus) -> Res<()> {
```
Is there a way to make places where we're calling `result::result` a little more clear? Can it be named `res_from_status` or `res_from_rv` or something similar?

agrover replied:
or can an `impl TryFrom<ssl::SECStatus> for Res` be of help?

martinthomson replied:
Given that `SECStatus` is always convertable, that would be `From`, wouldn't it?  Except that we don't define any of the involved types:
* `SECStatus` is a type alias for `i32`.
* `Res` is a type alias for `Result`.
* `From` is similarly in the standard library.

This seems like a great idea, but we can't realistically do that.  Renaming the function is the best we can do.  `res_from_secstatus` is the best I can come up with offhand.  I've opened #166.

agrover commented on neqo-crypto/src/aead.rs:
```diff
@@ -76,15 +76,14 @@ impl Aead {
             secret,
             p.as_ptr() as *const i8,
-            p.len() as u32,
+            p.len().try_into().unwrap(),
```
prefer tryfrom? e.g.  `u32::try_from(p.len()).unwrap()`. More explicit about target type.

martinthomson replied:
I think that I might prefer try_from where possible.  That doesn't seem to be idiomatic and I don't want to just hit this one instance when we do that.

martinthomson commented:
I've rebased this and done some more serious refactoring of the error handling pieces.  I've removed a couple of files in the process, which were created before I understood the whole conversion piece better.
